mpremote: Allow user configuration on Windows

Josverl opened the pull request:
Simplifies the use of mpremote's user configuration on Windows, by also considering the Windows standard environment variable `APPDATA`, which is the OS default for storing any user/application related file-based configuration. ( changed from USERPROFILE) 

This avoids the need for a Windows user to set the `HOME` or `XDG_CONFIG_HOME` environment variable, both of which are non-standard on Windows.

Per Microsoft documentation:

> The recommended location for a Windows application to store user settings is in the %APPDATA% folder. This folder is a hidden folder in the user's profile directory that is designed to store application-specific data, including user settings.

However Windows users also are used to %USERPROFILE%, which was established as a location in a previous century.
in order to adhere to both recommended usage ( $XDG_CONFIG_HOME and %APPDATA%) as well as common user practices ($HOME and %USERPROFILE%) , all variables used in a scan for a valid configuration file, and this first config file is used.

Simply stated : newer standards are preferred over the older standards, and Linux environment variables take precidence.

| # | ENV_VAR         | env_value                    | Path_checked                              | platform.system()              |  
| - | --------------- | ---------------------------- | ----------------------------------------- | ------------------------------ | 
| 1 | XDG_CONFIG_HOME | $HOME/**.config**            | /home/foo/**.config**/mpremote                | Linux, Windows(MSYS32), Darwin | 
| 2 | HOME            | /home/boo                    | /home/foo/**.config**/mpremote                | Linux, Windows(MSYS32), Darwin | 
| 3 | APPDATA         | C:\Users\foo\AppData\Roaming | C:\Users\foo\AppData\Roaming\mpremote | Windows                        | 
| 4 | USERPROFILE     | C:\Users\foo                 | C:\Users\foo\mpremote                     | Windows                        |

stinos commented on tools/mpremote/mpremote/main.py:
```diff
@@ -337,7 +337,9 @@ def load_user_config():
         path = os.getenv("HOME")
         if path is None:
-            return config
+            path = os.getenv("USERPROFILE")
```
By now it would in my opinion be nicer to rewrite this like
```
for var in ["XDG_CONFIG_HOME", "HOME", "USERPROFILE"]:
  path = os.getenv(var)
  if path is not None:
    break
```

Josverl replied:
makes sense 👍

Josverl commented on tools/mpremote/mpremote/main.py:
```diff
@@ -333,13 +348,15 @@ def load_user_config():
+        path = os.getenv(var)
+        if path:
+            path = os.path.join(path, ".config")
+            path = os.path.join(path, _PROG)
```
happy to simplify that as well.
also there is the case where the config file may contain utf-8 caracters 
![image](https://user-images.githubusercontent.com/981654/195064574-7f322644-387e-415f-acef-9296a6dc9310.png)
simple to fix : ` with open(config_file, encoding="utf-8") as f:`

dpgeorge commented on tools/mpremote/mpremote/main.py:
```diff
@@ -27,14 +28,14 @@
     do_resume,
+    do_run,
     do_soft_reset,
+    do_umount,
```
please don't do any reformatting when it's not related to the commit (reformatting can be done separately, at least in a separate commit)

Josverl replied:
IIRC the formatting was added by the recommended pre-commit hook, as apparently the original code was not up to spec.
makes sense to to seperate this out in a  commit, ( or I can remove this , but then the PR will fail again on CI)

dpgeorge commented on tools/mpremote/mpremote/main.py:
```diff
@@ -93,7 +94,9 @@ def _bool_flag(cmd_parser, name, short_name, default, description):
-        "device", nargs=1, help="Either list, auto, id:x, port:x, or any valid device name/path"
+        "device",
+        nargs=1,
+        help="Either list, auto, id:x, port:x, or any valid device name/path",
```
this looks like unnecessary reformatting?

dpgeorge commented on tools/mpremote/mpremote/main.py:
```diff
@@ -333,21 +347,21 @@ def load_user_config():
-    path = os.path.join(path, _PROG)
-    config_file = os.path.join(path, "config.py")
+        return config
+    config_file = os.path.join(path, ".config", _PROG, "config.py")
```
I'm not sure this logic of finding the path is correct.  Eg prior to this change we had:
1. if XDG_CONFIG_HOME is set then `path="<XDG_CONFIG_HOME>/mpremote"`
2. else if HOME is set then `path="<HOME>/.config/mpremote"`

With this patch, `path` is just `<XDG_CONFIG_HOME>` or `<HOME>`.

Josverl replied:
good catch, Ill rever to an earlier version that keeps the original behavior.
@stinos ,i folded the code based on your earlier review. I assume you agree with Damien ?

stinos replied:
Right, I misunderstood what it was doing. Could still use the loop which is imo a clearer way of expressing 'try these paths, with this priority' but then the loop items must be tuples like `[("XDG_CONFIG_HOME", "mpremote"), ("HOME", "config/mpremote"), ...]` which is in turn perhaps less clear.

Josverl replied:
- [x] I've updated the code and verified the behavior with a few tests before & after.
        
- [x] Formatting is removed (sorry for that, some tools try to be too clever) 
- [ ] I have a few tests in pytest, happy to add them somewhere , but could not find how / where test for the tools are located.

Josverl commented:
As this was not yet merged I made a change: 

* For *nix / Mac - leave the original logic and behaviour
* for Windows : use %APPDATA%/mpremote/config.py

stinos commented on tools/mpremote/mpremote/main.py:
```diff
@@ -317,31 +319,38 @@ def argparse_none(description):
-    if path is None:
-        path = os.getenv("HOME")
+    # use $XDG_CONFIG_HOME, if on Windows use $APPDATA
+    if os.name != "nt":
```
For a typical MSYS2 installation with a mingw-w64-x86_64-python3 installed, which is used a lot with MicroPython I think, and is what the non-msvc Windows port is built for, this is not ideal because its name is `nt` but it has a `HOME` directory and that is what you'd want to be used. There might be ways to detect that, but simpler is to just use `APPDATA` as fall back only `if not getenv('HOME') and `os.name == nt`.

Josverl replied:
Thanks for pointing that out. 
Its hard finding a 'weighted good' that works reliably and understandably across platforms. 
What are the changes of  such a 'MSYS2 installation with a mingw-w64-x86_64-python3' using "XDG_CONFIG_HOME" ?

stinos replied:
> using "XDG_CONFIG_HOME"

That's not a thing in standard MSYS2 as far as I can tell, i.e. if people want to use it they likely declare it themselves, but in any case: it would follow the same logic as there is now, so XDG_CONFIG_HOME taking precedence over HOME.

Josverl replied:
was already the care - Confirmed via tests for the the current PR

stinos commented on docs/reference/mpremote.rst:
```diff
@@ -428,8 +428,9 @@ Shortcuts can be defined using the macro system.  Built-in shortcuts are:
-located at ``.config/mpremote/config.py``. This file should define a
-dictionary named ``commands``. The keys of this dictionary are the shortcuts
+located at ``.config/mpremote/config.py`` relative to the ``XDG_CONFIG_HOME`` or ``HOME`` environment variable on unix systems
+, or on Windows relative to ``HOME``, ``USERPROFILE`` or ``APPDATA``.
```
That's not what the code seems to do: it tries all of "XDG_CONFIG_HOME", "HOME", "USERPROFILE", "APPDATA" irregardless of platforms?

stinos replied:
Yes this is just a detail. I meant this sentence makes it sound as if on unix the mpremote looks for HOME etc and on windows for USERPROFILE while in reality the code scans all variables both on unix and windows. Which is good enough in my opinion, and can be stated like 'Additional shortcuts can be defined by in the user configuration file, which is located at .config/mpremote/config.py relative to the XDG_CONFIG_HOME, HOME OR USERPROFILE environment variable (searched in that order).'

stinos commented on tools/mpremote/mpremote/main.py:
```diff
@@ -355,17 +357,23 @@ def load_user_config():
-            return config
-        path = os.path.join(path, ".config")
-    path = os.path.join(path, _PROG)
+    # use $XDG_CONFIG_HOME,$HOME $env:USERPROFILE% or $env:APPDATA
```
That's pretty clear from the code below so I'd leave out this comment, or else use something like 'Look for configuration in common directories'

stinos commented on tools/mpremote/mpremote/main.py:
```diff
@@ -355,17 +357,23 @@ def load_user_config():
-    path = os.path.join(path, _PROG)
+    # use $XDG_CONFIG_HOME,$HOME $env:USERPROFILE% or $env:APPDATA
+    path = None
+    for env_var in ("XDG_CONFIG_HOME", "HOME", "USERPROFILE", "APPDATA"):
```
Sorry didn't think this through earlier, but I think there's actually zero chance APPDATA is ever going to be used, since a windows system without USERPROFILE would be pretty much broken. Likewise any msys/cygwin like system will have HOME. So seems better to not use APPDATA at all.

Josverl replied:
The advantage of using APPDATA is that it roams across all my computers for free  ( no need for a .files repo ) 
so for my N=3 , there is a 100% change that it is used. 
USERPROFILE is checked first for those who prefer a old-school manual approach and will prefer a `%USERPROFILE%\.mpremote` folder over `%APPDATA%\.mpremote`

stinos replied:
So just to be clear: you do not have the USERPROFILE environment variable?

stinos commented on tools/mpremote/mpremote/main.py:
```diff
@@ -355,17 +357,23 @@ def load_user_config():
+            path = os.path.join(path, ".config", _PROG, "config.py")
+            break
+        elif os.path.exists(os.path.join(path, _PROG, "config.py")):
+            # Windows style
```
Not sure it's worth making this distinction, there's plenty of programs (including micropython itself) which just put dot directories in in the home directory on windows as well.

Josverl replied:
Agree that adding an additional layer of .config does not help, but that is the **current** behavior of mpremote, and removing that would result in a breaking change for all current *nix users that have a config.
That is fine with me - but probably something that would need to be called out in the change notes. 
The Freedektop specs are not that clear either to me 
https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
but that is the most likely source for the `.config` that I could pinpoint: "If $XDG_CONFIG_HOME is either not set or empty, a default equal to $HOME/.config should be used."

Also, based on your earlier feedback , there may be Windows users that use MSYS2/mingw, and expect to use a Unix convention on Windows, hence the check for Unix Style on Windows.

Josverl replied:
Perhaps @jimmo or @dpgeorge can help resolve the intended paths to use across the different client platforms.
I also found [a similar conversions on SO](https://stackoverflow.com/a/22161688/7909427) regarding which platforms use what paths for configuration.

currently my recommendation would be:

| # | ENV_VAR         | env_value                    | Path_checked                              | platform.system()              |  |
| - | --------------- | ---------------------------- | ----------------------------------------- | ------------------------------ | - |
| 1 | XDG_CONFIG_HOME | /home/foo/.config            | /home/foo/.config/mpremote                | Linux, Windows(MSYS32), Darwin |  |
| 2 | HOME            | /home/boo                    | /home/foo/.config/mpremote                | Linux, Windows(MSYS32), Darwin |  |
| 3 | USERPROFILE     | C:\Users\foo                 | C:\Users\foo\mpremote                     | Windows                        |  |
| 4 | APPDATA         | C:\Users\foo\AppData\Roaming | C:\Users\josverl\AppData\Roaming\mpremote | Windows                        |  |

That means that `.config` only will be added to `$HOME`, but please validate my understanding.
 
 Once we agree on what the outcome should be, I'll update the code accordingly

stinos replied:
> and removing that would result in a breaking change for all current *nix users that have a config

Sorry should have been more clear, I'm not saying the current behavior should change, just that as far as I'm concerned it's not needed to that extra location (`os.path.join(path, _PROG, "config.py")`) just for windows, but instead have it also use the same as *nix (`os.path.join(path, ".config", _PROG, "config.py")`)

Josverl replied:
It just does not make sense to enforce Linux naming conventions to Windows userprofiles.
I know I may be opiniated as one of these OSes is created by my current employer, but I stand by this design.

Josverl replied:
Switching to use [platformdirs](https://platformdirs.readthedocs.io/en/latest/index.html) to defer this to a party that has already solved the issue across platforms and configurations styles.

stinos commented on tools/mpremote/mpremote/main.py:
```diff
@@ -341,13 +325,15 @@ def argparse_none(description):
-            "command": "connect {}{}".format(port, port_num),
-            "help": 'connect to serial port "{}{}"'.format(port, port_num),
+        if port_num == 0 and port == "COM":
+            continue  # skip COM0 as it does not exist
```
mpremote.rst still mentions COM0 in the docs for the connect command, best to remove it there as well

Josverl commented:
Rebased again to simplify a merge.
Apparent Errors in CI/CD seem to be grounded into errors thrown by untouched parts the repo

Josverl commented:
rebased to allow merge

Josverl committed:
tools/mpremote: Add platformdirs dependency to requirements.txt.

Needed to easily find the user configuration file.

Signed-off-by: Jos Verlinde <Jos_Verlinde@hotmail.com>

Josverl committed:
tools/mpremote: Locate config.py location across different host OSes.

Use `platformdirs.user_config_dir()` (see
https://platformdirs.readthedocs.io/en/latest/api.html#user-config-directory)
to provide portability across many different OSes and configuration styles.

Signed-off-by: Jos Verlinde <Jos_Verlinde@hotmail.com>

Josverl committed:
docs/reference/mpremote: Document location of config file.

Signed-off-by: Jos Verlinde <Jos_Verlinde@hotmail.com>

dpgeorge reviewed (approved):
This is a really nice simplification and enhancement.  I usually don't like dependencies, but in this case IMO it's worth it, `platformdirs` brings in a lot of useful (and tricky) functionality.

Tested it locally (on Arch Linux) and it works well.  Arch has a `python-platformdirs` package (which I already had installed), and it continues to find my config in `$(HOME)/.config/mpremote/config.py`.

Very good!

dpgeorge commented:
No merged.

Sorry @Josverl this took so long to get through, but good that it's now much simpler with `platformdirs`.
